# PAPER_Hansen_ScientificReports_2023
## Recovery of the gut microbiome following enteric infection and persistence of antimicrobial resistance genes in specific microbial hosts

### Zoe A. Hansen, Karla Vasco, James T. Rudrik, Kim T. Scribner, Lixin Zhang, Shannon D. Manning

### Data
Raw data for this study are available on NCBI under the BioProjects PRJNA862908 and PRJNA660443 (BioSamples SAMN29999523 to SAMN29999673 and SAMN15958881 to SAMN15958950, respectively).

### Citation 


### Abstract
Enteric pathogens cause widespread foodborne illness and are increasingly resistant to important antibiotics yet their ecological impact on the gut microbiome and resistome is not fully understood. Herein, shotgun metagenome sequencing was applied to stool DNA from 60 patients (cases) during an enteric bacterial infection and after recovery (follow-ups). Overall, the case samples harbored more antimicrobial resistance genes (ARGs) with greater resistome diversity than the follow-up samples (p < 0.001), while follow-ups had more diverse gut microbiota (p < 0.001). Although cases were primarily defined by genera Escherichia, Salmonella, and Shigella along with ARGs for multi-compound and multidrug resistance, follow-ups had a greater abundance of Bacteroidetes and Firmicutes phyla and resistance genes for tetracyclines, macrolides, lincosamides, and streptogramins, and aminoglycosides. A host-tracking analysis revealed that Escherichia was the primary bacterial host of ARGs in both cases and follow-ups, with a greater abundance occurring during infection. Eleven distinct extended spectrum beta-lactamase (ESBL) genes were identified during infection, with some detectable upon recovery, highlighting the potential for gene transfer within the community. Because of the increasing incidence of disease caused by foodborne pathogens and their role in harboring and transferring resistance determinants, this study enhances our understanding of how enteric infections impact human gut ecology.

### Acknowledgements and Funding
This work was supported by the National Institutes of Health [grant number U19AI090872 to S.D.M. and J.T.R.]. Salary support was provided by the U.S. Department of Michigan Sequencing and Academic Partnerships for Public Health Innovation and Response (MI-SAPPHIRE) initiative at the MDHHS supported by the CDC through the Epidemiology and Laboratory Capacity for Prevention and Control of Emerging Infectious Diseases Enhancing Detection Expansion program (6NU50CK000510-02-07) (to S.D.M). Student support for Z.A.H. was provided by the College of Natural Science, the MSU Graduate School, and via the Ronald and Sharon Rogowski and Thomas S. Whittam Graduate Awards from the MSU Department of Microbiology and Molecular Genetics. We thank Ben Hutton and Jason Wholehan at the MDHHS for help with specimen processing and culture as well as Rebekah E. Sloup, Katherine Jernigan, and Samantha Carbonell, for help with community DNA isolation and sequencing. We also thank Dr. James M. Tiedje for the helpful discussions about these data.
